notes/force-move-implementation.md

```diff
@@ -164,9 +175,110 @@ Checking signatures is likely to be a signification part of the logic. Here's a
 
 **Can we optimize this further by changing the way states / sigs are passed in?** For example passing the arrays in last-state-first might help.
 
-## Updates call signatures
+## ForceMove interface
+
+With these considerations in mind, the ForceMove implementation should be something like the below.
+
+### Types and storage
+
+```javascript
+
+struct FixedPart {
+ string chainId;
+ address[] participants;
+ uint256 channelNonce;
+}
+
+struct VariablePart {
+  uint256 turnNum;
+  bool isFinal;
+}
+
+struct State { // participants sign this
+  uint256 turnNum;
+  bool isFinal;
+  bytes32 channelId;  // keccack(FixedPart)
+  bytes32 variablePartHash; //keccak(VariablePart)
+}
+
+struct ChannelStorage {
+  uint256 turnNumRecord;
+  uint256 finalizesAt;
+  bytes32 stateHash; // keccak(State)
+  address challengerAddress;
+}
+
+mapping(address => bytes32) public channelStorageHashes;
 
-[TODO] Rewrite the method call signatures to avoid any unnecessary repetition in the data being passed in.
+```
+
+### Public methods:
+
+```javascript
+
+
+function forceMove(uint turnNumRecord, FixedPart fixedPart, VariableParts[] variableParts, uint newTurnNumRecord, bool[] isFinals, Signature[] sigs, Signature challengerSig) public returns() {
```

Comment:
Currently this looks to be missing the `appDefinition`. If that was added to the `FixedPart` (but left out when calculating the `channelId`) that would fix it.

Reply:
Also, it might be more efficient to pass the `finalizationPoint` (as a `uint8`) - the index of the state where `isFinal` changes to `true`, instead of the `bool[] isFinals`.

Reply:
- It may not be more efficient, since that `uint8` will get converted to a `uint256` anyway (see [here](https://mudit.blog/solidity-gas-optimization-tips/)). We can get 32 bools in the same slot which is probably plenty.
- The gas saving would most likely be pretty small anyway (I believe 68 gas per byte of calldata)
- I'm not sure it's worth the extra pain for clients and the decrease in readability

Reply:
> Currently this looks to be missing the `appDefinition`. If that was added to the `FixedPart` (but left out when calculating the `channelId`) that would fix it.

Have addressed in 4de94bf